mergekit/scripts/extract_lora.py

```diff
@@ -0,0 +1,194 @@
+import json
+import os
+
+import bitsandbytes as bnb
+import click
+import torch
+from peft.tuners.lora import QuantLinear
+from safetensors.torch import save_file
+from tqdm import tqdm
+from transformers import AutoConfig, AutoModelForCausalLM
+
+from mergekit.common import ModelReference
+from mergekit.io import LazyTensorLoader
+from mergekit.options import add_merge_options
+
+
+def _low_rank_decomposition(weight, reduced_rank=16):
+    """
+    Decompose a 2D matrix into low-rank matrices A and B using SVD.a
+
+    :param weight: The matrix to decompose, of shape (H, W)
+    :param reduced_rank: The final rank of the decomposition
+    :return: A tuple of tensors (A, B)
+    """
+    if weight.dim() != 2:
+        raise ValueError(
+            f"Only support 2D matrix, but your input has {weight.dim()} dimensions."
+        )
+
+    # SVD Decomposition
+    U, S, Vh = torch.linalg.svd(weight, full_matrices=False)
+
+    # Truncated matrices
+    A = Vh[:reduced_rank, :]
+    B = U[:, :reduced_rank] @ torch.diag(S[:reduced_rank])
+
+    return A, B
+
+
+def decompose_delta_weight(new_weight, base_weight, reduced_rank, device=None):
+    if device is None:
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+
+    new_weight = new_weight.to(device)
+    base_weight = base_weight.to(device)
+
+    """
+    Decompose the delta weight into low-rank matrices A and B.
+
+    :param new_weight: The updated weight matrix after applying LoRA.
+    :param base_weight: The original weight matrix before LoRA.
+    :param reduced_rank: The rank for the low-rank decomposition.
+    :return: A tuple of tensors (A, B)
+    """
+    delta_weight = new_weight - base_weight
+
+    max_rank = min(delta_weight.shape)
+    assert (
+        reduced_rank <= max_rank
+    ), f"The specified rank ({reduced_rank}) must be smaller than or equal to the rank of the weight matrices ({max_rank})"
+
+    A, B = _low_rank_decomposition(delta_weight, reduced_rank=reduced_rank)
+
+    return A, B
+
+
+def find_all_linear_names(model):
+    cls = (bnb.nn.Linear4bit, bnb.nn.Linear8bitLt, torch.nn.Linear, QuantLinear)
+
+    names = []
+    for name, module in model.named_modules():
+        if (
+            isinstance(module, cls)
+            or "Linear" in module.__class__.__name__
+            and module.__class__.__name__ not in ("LlamaLinearScalingRotaryEmbedding",)
+        ):
+            names.append(name)
+
+    return names
+
+
+def get_linear_module_names(model_id):
+    model = AutoModelForCausalLM.from_pretrained(
+        model_id, state_dict={}, device_map="meta"
+    )  # avoid loading weights as we won't need them
```

Comment:
👍 Good approach for this. Someday we might put all of the info needed for this in ArchitectureInfo but this is perfect right now.